### PR TITLE
scheduler: fix a bug that happens when prewrite meets locks.

### PR DESCRIPTION
**Pull request by disksing**
If primary key is locked, other keys should not be submitted before primary key.

@siddontang @BusyJay @zhangjinpeng1987

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
LGTM @coocood PTAL

File: src/storage/txn/scheduler.rs [hunk 1]
```diff
@@ -400,16 +400,25 @@ fn process_write_impl(cid: u64,
...
+                    }
                     Err(e) => return Err(Error::from(e)),
                 }
             }
             let res = results.drain(..).map(|x| x.map_err(StorageError::from)).collect();
             let pr = ProcessResult::MultiRes { results: res };
-            (pr, txn.modifies())
+            if has_lock {
+                // Skip write stage if some keys are locked.
+                (pr, vec![])
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: When prewriting k1, k2 and k3,  k2 is locked, then results Ok(()), Err(locked), Ok(()) are returned. But k1, k3 are skipped, will they be retried when k2's lock is resolved.

Comment (disksing) [inline src/storage/txn/scheduler.rs]: Yes. Client resends the exact same prewrite request after locks are resolved.

File: src/storage/txn/scheduler.rs [hunk 2]
```diff
@@ -399,17 +399,25 @@ fn process_write_impl(cid: u64,
...
             let mut txn = MvccTxn::new(snapshot, start_ts, None);
-            let mut results = vec![];
+            let mut locks = vec![];
             for m in mutations {
                 match txn.prewrite(m.clone(), primary, options) {
-                    Ok(_) => results.push(Ok(())),
-                    e @ Err(MvccError::KeyIsLocked { .. }) => results.push(e.map_err(Error::from)),
+                    Ok(_) => {}
+                    e @ Err(MvccError::KeyIsLocked { .. }) => {
+                        locks.push(e.map_err(Error::from));
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: We can do the translation here instead of L417.

File: tests/storage/test_storage.rs [hunk 3]
```diff
@@ -131,6 +131,27 @@ impl AssertionStorage {
...
 
+    fn prewrite_locked(&self,
+                       mutations: Vec<Mutation>,
+                       primary: &[u8],
+                       start_ts: u64,
+                       expect_locks: Vec<(&[u8], &[u8], u64)>) {
+        let res = self.0.prewrite(Context::new(), mutations, primary.to_vec(), start_ts).unwrap();
+        let locks: Vec<(&[u8], &[u8], u64)> = res.iter()
+            .filter_map(|x| {
+                if let Err(storage::Error::Txn(txn::Error::Mvcc(mvcc::Error::KeyIsLocked { ref key,
```
Comment (BusyJay) [inline tests/storage/test_storage.rs]: Amazing!

File: tests/storage/test_storage.rs [hunk 4]
```diff
@@ -131,6 +131,27 @@ impl AssertionStorage {
...
+                                                                              ref primary,
+                                                                              ts,
+                                                                              .. }))) = *x {
+                    Some((key.as_ref(), primary.as_ref(), ts))
```
Comment (BusyJay) [inline tests/storage/test_storage.rs]: I don't think `as_ref` is necessary.

File: tests/storage/test_storage.rs [hunk 5]
```diff
@@ -578,6 +599,31 @@ fn test_txn_store_rawkv() {
...
+#[test]
+fn test_txn_store_lock_primary() {
+    let store = new_assertion_storage();
+    // txn1 locks "p" then aborts.
+    store.prewrite_ok(vec![Mutation::Put((make_key(b"p"), b"p1".to_vec()))],
+                      b"p",
+                      1);
+
+    // txn2 wants to write "p", "s".
+    store.prewrite_locked(vec![Mutation::Put((make_key(b"p"), b"p2".to_vec())),
```
Comment (BusyJay) [inline tests/storage/test_storage.rs]: We can prewrite_ok "s2" after this line to verify other keys won't be wrtten if primary key is locked.

Comment (disksing) [inline tests/storage/test_storage.rs]: Do you mean start a new txn to write only "s"?
It's ok for other txns to write "s" when it's not locked.

File: tests/storage/test_storage.rs (hunk 4 above)
Comment (disksing) [inline tests/storage/test_storage.rs]: `key` and `primary` are `&Vec<u8>` and we need `&[u8]`.

File: tests/storage/test_storage.rs (hunk 5 above)
Comment (disksing) [inline tests/storage/test_storage.rs]: txn3 checks it by write both "p" and "s".

**Comment by BusyJay** [comment]
LGTM

**Comment by disksing** [comment]
@hhkbp2 PTAL

File: tests/storage/test_storage.rs (hunk 3 above)
Comment (siddontang) [inline tests/storage/test_storage.rs]: Why amazing here :sob:

Comment (BusyJay) [inline tests/storage/test_storage.rs]: The match pattern is amazingly long!

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**